### PR TITLE
bpo-40636: PEP 618 implementation

### PR DESCRIPTION
**Pull request by gvanrossum**
https://bugs.python.org/issue40636

### CODE REVIEW

File: Python/bltinmodule.c
Comment (gvanrossum) [inline Python/bltinmodule.c]: It bothers me that the implementation gives different errors depending on whether it's the first iterator that's longer or any of the later ones. It seems to assume that the first iterator is "right" and the others are either too short or too long. But isn't it more the case that all we know is that they aren't all of the same length -- we shouldn't judge which one is too long or too short. Perhaps the errors could be changed to something like "iterator N is shorter/longer than the preceding iterators"?

Comment (brandtbucher) [reply Python/bltinmodule.c]: Yeah, perhaps these are a bit too terse.  I'll think on some other options (I also agree that unifying the messages is probably a good idea).

**Comment by gvanrossum** [comment]
I removed the "do not merge" from the title because we don't need three ways to indicate this -- two ways (draft mode and the do-not-merge label) seen quite sufficient.

Also one reason not to merge is now gone -- I accepted the PEP.

**Comment by gvanrossum** [comment]
Oh, you do need to add a bpo issue number and a news blurb.

File: Python/bltinmodule.c
Comment (vstinner) [inline Python/bltinmodule.c:2731]: You should update the prototype: `zip(*iterables)` => `zip(*iterables, strict=False)`.

**Review by vstinner — CHANGES_REQUESTED** [review]
* A new section is needed in https://docs.python.org/dev/whatsnew/3.10.html doc.
* https://docs.python.org/dev/library/functions.html#zip should be updated: `.. versionchanged:: 3.10`` markup is needed to describe what changed.

Note: I'm kind of confused. @gvanrossum created the PR, but @brandtbucher implemented it?

**Comment by brandtbucher** [comment]
> * A new section is needed in https://docs.python.org/dev/whatsnew/3.10.html doc.
> * https://docs.python.org/dev/library/functions.html#zip should be updated: ` ``.. versionchanged:: 3.10`` ` markup is needed to describe what changed.

@cool-RR volunteered to take the lead on documentation and testing, which can probably be made in separate PRs. I'm happy to monitor the BPO issue until everything is done.

Ram: be sure to include some round-trip `pickle` tests for `zip(..., strict=True)` as well. We didn't discuss it before, but it's a decent chunk of what changed here.

> Note: I'm kind of confused. @gvanrossum created the PR, but @brandtbucher implemented it?

He opened the PR from my branch in order to provide review comments on it. I was confused at first, too!

**Comment by brandtbucher** [comment]
@gvanrossum it looks like I don't have permissions to change the title of the PR or un-draft it. Can you do so, linking bpo-40636?

**Comment by gvanrossum** [comment]
Sorry about the unusual workflow. Next time I'll just ask the author to create the PR.

**Comment by vstinner** [comment]
Well, if there are permission issues, maybe @brandtbucher can create a new PR.

**Comment by cool-RR** [comment]
I'm very happy that Guido accepted the PEP :)

I'm responsible for writing the documentation and tests. If anyone has issues or requests regarding these please let me know. 

@brandtbucher I'm trying to figure out the best way to manage this collaboration. You suggested that I open separate PRs, but would that also mean separate BPO issues? Will I start my branch from your branch and then rebase?

**Comment by brandtbucher** [comment]
@vstinner, I believe I can still push to the branch. If any other part of the workflow is disrupted, I'll just go ahead and open a new one.

@cool-RR You can use the same BPO issue. Hold off until this is merged (won't be long), then just branch off of `master` once it's in like normal.

**Comment by gvanrossum** [comment]
@cool-RR There's not even a reason to wait until this PR has been merged, since you're not editing the same files (presumably). The only thing that needs to happen is that *merging* your PR should come after Brandt's PR is merged. But you can request review immediately.

**Comment by brandtbucher** [comment]
That's true, docs could start now.

**Comment by brandtbucher** [comment]
@gvanrossum, I'm not able to request a review from you. But if you and Victor approve, then this is probably ready.

```py
>>> list(zip("ab", "cde", strict=True))
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
ValueError: zip() argument 2 is longer than argument 1
```

```py
>>> list(zip("abc", "def", "ghi", "jk", strict=True))
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
ValueError: zip() argument 4 is shorter than arguments 1-3
```

**Review by brandtbucher — APPROVED** [review]

File: Python/bltinmodule.c
Comment (vstinner) [inline Python/bltinmodule.c]: Why not using a separated member for strict? I expected "int strict".

File: Misc/NEWS.d/next/Core and Builtins/2020-06-17-10-27-17.bpo-40636.MYaCIe.rst
Comment (gvanrossum) [inline Misc/NEWS.d/next/Core and Builtins/2020-06-17-10-27-17.bpo-40636.MYaCIe.rst]: The "Patch by ..." phrase usually isn't a separate paragraph.
```suggestion
:exc:`ValueError` if the arguments are exhausted at differing lengths. Patch by Brandt Bucher.
```

File: Python/bltinmodule.c
Comment (gvanrossum) [reply Python/bltinmodule.c]: Agreed. Encoding an extra bit in the sign is so 1975. :-)

File: Python/bltinmodule.c
Comment (gvanrossum) [inline Python/bltinmodule.c]: When is tp_name ever not "zip"? I suppose when the user subclasses it. But at that point we don't know anything about the argument order (the subclass could have a different constructor signature).

File: Lib/test/test_builtin.py
Comment (gvanrossum) [inline Lib/test/test_builtin.py:1674]: I'd like to see a test for pickling and unpickling of a zip(strict=True) object.

File: Python/bltinmodule.c
Comment (vstinner) [reply Python/bltinmodule.c]: But this encoding (size + strict) makes sense for getstate/setstate: it provides backward compatibility for pickle which is nice!

Comment (brandtbucher) [reply Python/bltinmodule.c]: I think I was feeling a bit too clever here... and preempting the argument that I was bloating `zip` instances. ;)

Changes coming soon.

File: Lib/test/test_builtin.py
Comment (brandtbucher) [reply Lib/test/test_builtin.py:1674]: I'm sure @cool-RR will include tests for this in their follow-up PR. Would you like me to add one here, though?

Comment (gvanrossum) [reply Lib/test/test_builtin.py:1674]: Yes, esp. given Victor's observation about pickling.

File: Lib/test/test_builtin.py
Comment (gvanrossum) [inline Lib/test/test_builtin.py]: Let me propose another pickle test. Let's take a zip() iterator, pickle it under Python 3.9, and save the string. Then check that it can still be unpickled correctly here. Also check that pickling the same zip() iterator produces the same string under 3.10 as it did under 3.9. (Or perhaps the latter test is sufficient, since we already know that it roundtrips.) Bonus points for doing this for all pickle protocols supported by 3.9.

Comment (brandtbucher) [reply Lib/test/test_builtin.py]: Interesting idea! I'll have that up tonight.

Comment (brandtbucher) [reply Lib/test/test_builtin.py]: Although, a test like this *does* effectively lock in our pickles going forward (and not just for zip, but also tuples / tuple iterators / ints / whatever we zip in the test). I can do an ad-hoc verification that it works, but are we sure it needs a unit test that makes sure all future zip pickles match those from 3.9?

**Comment by gvanrossum** [comment]
Yes, pickles need to be interoperable between versions. People store pickles in databases.

File: Lib/test/test_builtin.py
Comment (brandtbucher) [reply Lib/test/test_builtin.py]: Done (added the same test for strict pickles as well).

**Review by gvanrossum — COMMENTED** [review]
LGTM. I can't actually approve. @vstinner ?

**Comment by brandtbucher** [comment]
Don't worry @gvanrossum, I did for you! :upside_down_face:

File: Lib/test/test_builtin.py
Comment (vstinner) [inline Lib/test/test_builtin.py:1673]: Rather than testing iterables once zip() failed, would it be possible to write an helper function which iterate on zip, store items in a list, and catch ValueError?

Something like:

```
# def consume_zip_strict(self, iterables):
items = []
with self.assertRaises(ValueError):
    for item in zip(*iterables, strict=strict):
        items.append(item)
return items
```

The idea is to ensure that we get the expected partial list.

File: Misc/NEWS.d/next/Core and Builtins/2020-06-17-10-27-17.bpo-40636.MYaCIe.rst
Comment (vstinner) [inline Misc/NEWS.d/next/Core and Builtins/2020-06-17-10-27-17.bpo-40636.MYaCIe.rst]: (English is not my first language.) Is "differing" correct here?

```suggestion
:exc:`ValueError` if the arguments are exhausted at different lengths.
```

File: Python/bltinmodule.c
Comment (vstinner) [inline Python/bltinmodule.c]: The code in the error label is non-trivial. You may add comments. Here I propose:

```
// next() on argument i raised an exception (not StopIteration)
```

File: Python/bltinmodule.c
Comment (vstinner) [inline Python/bltinmodule.c:2664]: I suggest to always call PyErr_Clear(); here, to clear the current exception, if StopIteration was raised.

Otherwise, tp_iternext is called below with an exception set. tp_iternext can return NULL without setting an exception, but it would "inherit" the exception which was already set.

Remove the PyErr_Clear() below (in `if (i)` block).

File: Python/bltinmodule.c
Comment (vstinner) [inline Python/bltinmodule.c]: This error path is not tested: would you mind to write an unit test for it?

File: Python/bltinmodule.c
Comment (vstinner) [inline Python/bltinmodule.c]: This errror path is not tested. Would you mind to add two unit tests for it?

* One test where an iterator raises StopIteration
* One test where an iterator raises an exception other than StopIteration

... This is bug in this code: if StopIteration is raised, ValueError is not raised whereas a following argument can be not exhausted yet.

zip(a, b,c, strict=True): if a is exhausted, b raises StopIteration, but c is not exhausted => StopIteration is passed through, whereas a ValueError must be raised because c is not exhausted.

File: Python/bltinmodule.c
Comment (vstinner) [inline Python/bltinmodule.c:2689]: Please add a comment to explain this code path. I understand that:


```
// The argument i is exhausted
```

File: Python/bltinmodule.c
Comment (vstinner) [inline Python/bltinmodule.c:2691]: Please add a comment to explain this code path. I understand that:

```
// All arguments are exhausted successfully
```

File: Lib/test/test_builtin.py
Comment (cool-RR) [reply Lib/test/test_builtin.py:1673]: Sounds good. @brandtbucher How are we handling the collaboration here? Should I be pushing commits into this branch, or do the tests on a separate PR, or would you like to take over this issue?

Another idea I thought of for the test, is to make the iterables passes to zip be of a special iterator class that has a global log for all the `__next__` actions done on them. At the end of the call, we could assert that the log is equal to exactly the correct order.

**Comment by gvanrossum** [comment]
(...and that's why I am not allowed to approve this PR. :-)

File: Misc/NEWS.d/next/Core and Builtins/2020-06-17-10-27-17.bpo-40636.MYaCIe.rst
Comment (brandtbucher) [reply Misc/NEWS.d/next/Core and Builtins/2020-06-17-10-27-17.bpo-40636.MYaCIe.rst]: Pretty sure "differing" is correct here (I think a strict reading of "different" may mean that *no two lengths* were the same).

I'll just change it anyways, since "different" probably makes more sense to more people.

File: Python/bltinmodule.c
Comment (brandtbucher) [reply Python/bltinmodule.c]: I always trip on the fact that Python code raises `StopIteration` but C code usually doesn't. Nice catch.

File: Python/bltinmodule.c
Comment (brandtbucher) [reply Python/bltinmodule.c:2664]: Yeah, my new test cases were causing asserts to be tripped up because of the unhandled exception, so this *needs* to be done.

File: Lib/test/test_builtin.py
Comment (brandtbucher) [reply Lib/test/test_builtin.py:1673]: @cool-RR, perhaps I could just address the comments and push the tests required for this review? They're mostly error handling for some specific edge cases, and I needed to write the tests anyways to diagnose problems with how I was handling them.

Then you could do your follow-up PR which checks the rest of the behavior, including all of the "normal" cases.

This is already a weird PR (I don't "own" it), so I worry that having different people pushing and addressing review comments would just create more problems / permission issues. Sorry...

**Comment by brandtbucher** [comment]
Thanks @vstinner, I think I addressed all of your comments.

The macOS `test_resource_tracker_reused` failure is unreleated (it's currently failing on other PRs).

File: Lib/test/test_builtin.py
Comment (cool-RR) [reply Lib/test/test_builtin.py:1673]: Sounds good.

File: Lib/test/test_builtin.py
Comment (vstinner) [inline Lib/test/test_builtin.py]: Please use a name longer than a single letter. MyError; Bug, etc. Whatever works :-)

Ditto for "class I:": Iterator, MyIterator, BuggyIter, etc.

File: Python/bltinmodule.c
Comment (vstinner) [inline Python/bltinmodule.c]:
```suggestion
        const char* plural = i == 1 ? " " : "s 1-";
```

File: Lib/test/test_builtin.py
Comment (vstinner) [inline Lib/test/test_builtin.py]: Can you check l1 immediately? Currently, it's not trivial to see the relationship between l1 and the expected value. Maybe write `assertEqual(self.iter..., [...])` (on multiple lines).

**Review by vstinner — COMMENTED** [review]
Thanks, the update test suite seems to have a better code coverage ;-)

**Review by vstinner — APPROVED** [review]
LGTM. Thanks for the better code coverage and additional comments.

I also checked for refleak: `./python -m test test_builtin -R 3:3` pass successfully (but I had to disable PtyTests, it's unrelated to this PR.)

**Comment by brandtbucher** [comment]
Thanks!

**Comment by cool-RR** [comment]
Congrats on the merge. If there are any more tests that you'd like me to write, let me know, with details.

**Comment by cool-RR** [comment]
Do you think we should look for places where `zip` is used in the stdlib that could benefit from `strict=True`? If I remember correctly there's at least one example in the PEP.

**Comment by gvanrossum** [comment]
I would be very conservative with such changes.

**Pull request merged**
